### PR TITLE
[ffmpeg/opencv4] Update opencv4 feature dependency and remove ffmpeg feature 'postproc' from default feature

### LINKED ISSUE #19334: current opencv[ffmpeg] feature makes the whole opencv library to be GPL licensed?
FFmpeg uses LGPL license by default which is pretty permissive when it comes to building a commercial software that uses ffmpeg as its part. 
This gives a good summary:
https://www.ffmpeg.org/legal.html
This is helpfull for whoever is going to take care of this issue, it provides a list of parts in ffmpeg which require GPL license:
https://chromium.googlesource.com/chromium/third_party/ffmpeg/+/refs/heads/master/LICENSE.md

Unfortunately when one builds opencv[ffmpeg], ffmpeg package by default is configured to be with GPL license. I am pretty sure it is because with --enable-gpl more filters are included which makes ffmpeg more usable.
But here it comes:
https://www.gnu.org/licenses/gpl-faq.html#IfLibraryIsGPL 
Anything linked against GPL becomes itself a GPL. And so on till the user's code itself. Which is absolutely not the thing that commercial developer wants to have. And being sure that opencv itself has always BSD or Apache2.0 license could be quite an illusion as we see from the case above.

I suggest changing ffmpeg configuration when user does `vcpkg install opencv[ffmpeg]` to be non-GPL by default.
How can I manually change ffmpeg package configuration that opencv defines?

I would say there could be a nice feature to check resulting licensing for every package installed based on individual licenses of dependant packages.

### PR DESCRIPTION
**Pull request by PhoebeHui**
Fixes https://github.com/microsoft/vcpkg/issues/19334

Changes:
1. Opencv4, opencv3 and opencv2 only requires feature ffmpeg[avcodec,avdevice,avformat,swresample,swscale] by default see https://github.com/opencv/opencv/blob/master/CMakeLists.txt#L1369, so we can update opencv4 to depend on these features instead
2. Remove the 'postproc' from ffmpeg default feature: ffmpeg default feature 'postproc' depends on ffmpeg[gpl], so the gpl would be installed, that cause ffmpeg is configured to be with GPL license by default, See details in issue https://github.com/microsoft/vcpkg/issues/19334

### CODE REVIEW

**Review by JackBoosY — APPROVED** [review]

**Comment by cenit** [comment]
why not also to opencv2?

**Comment by PhoebeHui** [comment]
@cenit, thanks for your review! The opencv2 has been updated.

**Comment by PhoebeHui** [comment]
opecv4:x64-uwp and opencv4:arm-uwp failed with following failures:

```
D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(266,9): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(267,9): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(275,13): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(341,45): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(397,38): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(420,5): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(420,27): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(436,9): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(443,9): error C4996: 'tiff_dummy_namespace::uint16': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(463,9): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(481,71): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(527,33): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(531,33): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(569,33): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]
         D:\buildtrees\opencv4\src\4.5.2-755f235ba0.clean\modules\imgcodecs\src\grfmt_tiff.cpp(573,33): error C4996: 'tiff_dummy_namespace::uint32': libtiff type deprecated; please use corresponding C99 stdint.h type [D:\buildtrees\opencv4\x64-uwp-dbg\modules\imgcodecs\opencv_imgcodecs.vcxproj]

```

**Comment by cenit** [comment]
@PhoebeHui is this valid also for  #18827? if so, it might be a tiff recent update which broke downstream ports using it, because it was not a problem when that pr was completed

**Comment by dg0yt** [comment]
Tiff change its type system with 4.3.0 (to standard types!). I explicitly stated this in the update (#18187). That PR also include a fix to the single port which had the same issue: https://github.com/microsoft/vcpkg/pull/18187/files#diff-9693267cc3a0c0a25b5e19b706db8f4eee2ad2758d8645e5f3e3438ca98105bd

Summary:
~~~
+    if(MSVC)
+        target_compile_options(foo PRIVATE /wd4996)
+    endif()
~~~

File: ports/opencv2/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
+{
+  "name": "opencv2",
+  "version": "2.4.13.4",
```
Comment (JackBoosY) [inline ports/opencv2/vcpkg.json:3]:
```suggestion
  "version": "2.4.13.7",
```

File: versions/baseline.json [hunk 2]
```diff
@@ -4617,16 +4617,16 @@
       "port-version": 0
     },
     "opencv2": {
-      "baseline": "2.4.13.7",
-      "port-version": 5
+      "baseline": "2.4.13.4",
```
Comment (JackBoosY) [inline versions/baseline.json:4620]:
```suggestion
      "baseline": "2.4.13.7",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by cenit** [comment]
again, just to be sure, isn't the same patch for tiff required also by opencv2 and not just opencv3 and opencv4?

**Comment by cenit** [comment]
yes, sorry you're right. This is a problem specific to uwp (which i still have to investigate because it looks suspicious it is found only on uwp).
Might just be that warnings are treated as errors in uwp, and the patch in fact just disables that warning, now that i think about it...

**Comment by cenit** [comment]
i was referring to the tiff uwp patch, but yes, opencv2 has other problems too on uwp, so no reason to patch it for now (and also, i don't know anyone using opencv2 on uwp, so no need to make it work in future too)

**Comment by PhoebeHui** [comment]
@BillyONeal, could you help merge this PR?

**Comment by cenit** [comment]
can you please add the fix in this comment to this PR?

https://github.com/microsoft/vcpkg/issues/19497#issuecomment-897771735

**Comment by BillyONeal** [comment]
> can you please add the fix in this comment to this PR? @PhoebeHui
> 
> [#19497 (comment)](https://github.com/microsoft/vcpkg/issues/19497#issuecomment-897771735)

I'm not sure I understand, that comment seems to indicate that there's a broken hash somewhere but I tried this locally without asset caching and it passed (as in the download succeeded and hash check passed)

@cenit  Do you want to block this change until that is applied? Thanks!

**Comment by cenit** [comment]
cuda feature had a hash change, it would be nice to archive opencv 4.5.2 with a proper hash before going to 4.5.3

**Comment by PhoebeHui** [comment]
@cenit, I tried it locally, and it looks the hash is correct.

```
[DEBUG] Downloading https://github.com/NVIDIA/NVIDIAOpticalFlowSDK/archive/79c6cee80a2df9a196f20afd6b598a9810964c32.zip
Error: Failed to download from mirror set:
File does not have the expected hash:
             url : [ https://github.com/NVIDIA/NVIDIAOpticalFlowSDK/archive/79c6cee80a2df9a196f20afd6b598a9810964c32.zip ]
       File path : [ E:\vcpkg\PhoebeHui\vcpkg\downloads\opencv-cache\nvidia_optical_flow\ca5acedee6cb45d0ec610a6732de5c15-79c6cee80a2df9a196f20afd6b598a9810964c32.zip.23488.part ]
   Expected hash : [ 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ]
     Actual hash : [ d80cdedec588dafaad4ebb8615349f842ecdc64d3ca9480fee7086d606e6f2362606a9a2ce273c5cf507be2840ec24bbcbe32c2962672c3bcfb72d31428ef73d ]
```

**Comment by PhoebeHui** [comment]
Updated! thanks @BillyONeal @dg0yt @cenit for your review!

**Comment by BillyONeal** [comment]
Thanks for your help folks!

**Pull request merged**